flink-scala/src/main/scala/org/apache/flink/api/scala/typeutils/ScalaEnumSerializerSnapshot.scala

```diff
@@ -0,0 +1,136 @@
+    this.previousEnumConstants = Preconditions.checkNotNull(previousEnumConstants)
+  }
+
+  override def getCurrentVersion: Int = ScalaEnumSerializerSnapshot.VERSION
+
+  override def writeSnapshot(out: DataOutputView): Unit = {
+    val outViewWrapper = new DataOutputViewStream(out)
+    try {
+      out.writeUTF(enumClass.getName)
+
+      out.writeInt(previousEnumConstants.length)
+      for ((name, idx) <- previousEnumConstants) {
+        out.writeUTF(name)
+        out.writeInt(idx)
+      }
+    } finally if (outViewWrapper != null) outViewWrapper.close()
+  }
+
+  override def readSnapshot(
+    readVersion: Int, in: DataInputView, userCodeClassLoader: ClassLoader): Unit = {
+    val inViewWrapper = new DataInputViewStream(in)
+    try {
+      if (readVersion == 1) {
+        enumClass = InstantiationUtil.deserializeObject(
+          inViewWrapper, userCodeClassLoader)
+
+        // read null from input stream
+        InstantiationUtil.deserializeObject(inViewWrapper, userCodeClassLoader)
+        previousEnumConstants = List()
+      } else if (readVersion >= 2) {
+        enumClass = Class.forName(
+          in.readUTF(), true, userCodeClassLoader).asInstanceOf[Class[E]]
+
+        val length = in.readInt()
+        val listBuffer = ListBuffer[(String, Int)]()
+
+        for (_ <- 0 until length) {
+          val name = in.readUTF()
+          val idx = in.readInt()
+          listBuffer += ((name, idx))
+        }
+
+        previousEnumConstants = listBuffer.toList
+      } else {
+        throw new IOException(
+          s"Cannot deserialize ${getClass.getSimpleName} with version $readVersion.")
+      }
+    } catch {
+      case e: ClassNotFoundException =>
+        throw new IOException("The requested enum class cannot be found in classpath.", e)
+    }
+    finally if (inViewWrapper != null) inViewWrapper.close()
+  }
+
+  override def restoreSerializer(): TypeSerializer[E#Value] = {
+    enumClass.newInstance().asInstanceOf[TypeSerializer[E#Value]]
+  }
+
+  override def resolveSchemaCompatibility(
+    newSerializer: TypeSerializer[E#Value]): TypeSerializerSchemaCompatibility[E#Value] = {
+    newSerializer match {
+      case newEnumSerializer: EnumValueSerializer[_] => {
+        if (enumClass.equals(newEnumSerializer.enum.getClass)) {
+          for ((previousEnumConstant, idx) <- previousEnumConstants) {
+            val enumValue = try {
+              newEnumSerializer.enum(idx)
+            } catch {
+              case _: NoSuchElementException =>
+                // couldn't find an enum value for the given index
+                return TypeSerializerSchemaCompatibility.incompatible()
+            }
+
+            if (!previousEnumConstant.equals(enumValue.toString)) {
+              // compatible only if new enum constants are only appended,
+              // and original constants must be in the exact same order
+              return TypeSerializerSchemaCompatibility.compatibleAfterMigration()
```

Comment:
I know that the previous checking code in `EnumValueSerializer` also returned _compatible after migration_ here but I don't think we do that migration here, so what happens if the index of the value actually changed? Maybe we should be conservative here and disallow any changes to the enum.

@igalshilman, what do you think about this?

Reply:
If the index of the value actually changed, there should be incompatible, I'll update it. because in `EnumValueSerializer#serializer` we use `Enumeration#Value.id`